framework/src/headers/ip/v6/icmp/ndp_options.rs

```diff
@@ -0,0 +1,157 @@
+use std::fmt;
+use std::slice;
+
+use num::FromPrimitive;
+
+use headers::ip::v6::icmp::ndp::NdpMessageContents;
+use headers::mac::MacAddress;
+
+#[derive(FromPrimitive, Debug, PartialEq, Hash, Eq, Clone, Copy)]
+#[repr(u8)]
+pub enum NdpOptionType {
+    SourceLinkLayerAddress = 1,
+    TargetLinkLayerAddress = 2,
+    PrefixInformation = 3,
+    RedirectHeader = 4,
+    MTU = 5,
+}
+
+impl fmt::Display for NdpOptionType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            NdpOptionType::SourceLinkLayerAddress => write!(f, "Source Link Layer Address"),
+            NdpOptionType::TargetLinkLayerAddress => write!(f, "Target Link Layer Address"),
+            NdpOptionType::PrefixInformation => write!(f, "Prefix Information"),
+            NdpOptionType::RedirectHeader => write!(f, "Redirect Header"),
+            NdpOptionType::MTU => write!(f, "MTU"),
+        }
+    }
+}
+
+
+/**
+    MTU NDP Option type
+
+       0                   1                   2                   3
+       0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |     Type      |    Length     |           Reserved            |
+      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+      |                              MTU                              |
+      +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+   Fields:
+
+      Type           5
+
+      Length         1
+
+      Reserved       This field is unused.  It MUST be initialized to
+                     zero by the sender and MUST be ignored by the
+                     receiver.
+
+      MTU            32-bit unsigned integer.  The recommended MTU for
+                     the link.
+
+   Description
+                     The MTU option is used in Router Advertisement
+                     messages to ensure that all nodes on a link use the
+                     same MTU value in those cases where the link MTU is
+                     not well known.
+
+                     This option MUST be silently ignored for other
+                     Neighbor Discovery messages.
+
+                     In configurations in which heterogeneous
+                     technologies are bridged together, the maximum
+                     supported MTU may differ from one segment to
+                     another.  If the bridges do not generate ICMP
+                     Packet Too Big messages, communicating nodes will
+                     be unable to use Path MTU to dynamically determine
+                     the appropriate MTU on a per-neighbor basis.  In
+                     such cases, routers can be configured to use the
+                     MTU option to specify the maximum MTU value that is
+                     supported by all segments.
+*/
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct MtuOption {
+    reserved: u16,
+    mtu: u32,
+}
```

Comment:
I think one short coming here is when we need to create a NDP packet (like for neighbor advertisement), we don't have a complete option struct with the type and length.

Reply:
this is correct.  The structs definitely have not been designed to accommodate writes

Reply:
Perhaps we should actually fully flush out each option type then?  It is easy enough to parse. 

Then we would add `option_type: u8` and `option_length: u8` to this struct, and create a struct for `SourceLinkLayerAddressOption` and such?

That would at least allow you to write it.

Reply:
@pauljamescleary and i chatted about having the type and length. we could have a separate struct like so ```pub struct NDPOptionMeta<T> {
    option_type: u8,
    option_length: u8,
    _parent: PhantomData<T>,
}``` and reference that on each option

Reply:
yeah I think we would want each option fully flushed out.

Reply:
I think this is resolved now @drunkirishcoder